Sources/NIO/Socket.swift

```diff
@@ -201,6 +201,45 @@ typealias IOVector = iovec
         }
     }
 
+    /// Receive data from the socket, along with aditional control information.
+    ///
+    /// - parameters:
+    ///     - pointer: The pointer (and size) to the storage into which the data should be read.
+    ///     - storage: The address from which the data was received
+    ///     - storageLen: The size of the storage itself.
+    ///     - controlBytes: A region of a buffer into which control data can be written. This parameter will be modified on return to be
+    ///         the slice of the data actually written into, if any.
+    /// - returns: The `IOResult` which indicates how much data could be received and if the operation returned before all the data could be received
+    ///     (because the socket is in non-blocking mode)
+    /// - throws: An `IOError` if the operation failed.
+    func recvmsg(pointer: UnsafeMutableRawBufferPointer, storage: inout sockaddr_storage, storageLen: inout socklen_t, controlBytes: inout Slice<UnsafeMutableRawBufferPointer>) throws -> IOResult<Int> {
+        var vec = iovec(iov_base: pointer.baseAddress, iov_len: pointer.count)
+        let localControlBytePointer = UnsafeMutableRawBufferPointer(rebasing: controlBytes)
+
+        return try withUnsafeMutablePointer(to: &vec) { vecPtr in
+            return try storage.withMutableSockAddr { (sockaddrPtr, _) in
+                var messageHeader = msghdr(msg_name: sockaddrPtr,
+                                           msg_namelen: storageLen,
+                                           msg_iov: vecPtr,
+                                           msg_iovlen: 1,
+                                           msg_control: localControlBytePointer.baseAddress,
+                                           msg_controllen: .init(localControlBytePointer.count),
+                                           msg_flags: 0)
+                defer {
+                    // We need to write back the length of the message and the control bytes.
```

Comment:
Should we still write these back if we throw?

Reply:
Yes. We only throw _after_ the syscall occurs, and the syscall may well have set the values we're reading. To clarify that, we ensure that the values are written back. Note that if these values were _not_ set, they'll be defaulted to the original values: `messageHeader.msg_namelen` is initialised above to `storageLen`, and `messageHeader.msg_controllen` is initialised to `localControlBytePointer.count`.